frame/bags-list/src/list/mod.rs

```diff
@@ -0,0 +1,774 @@
+// limitations under the License.
+
+//! Implementation of a "bags list": a semi-sorted list where ordering granularity is dictated by
+//! configurable thresholds that delineate the boundaries of bags. It uses a pattern of composite
+//! data structures, where multiple storage items are masked by one outer API. See [`ListNodes`],
+//! [`CounterForListNodes`] and [`ListBags`] for more information.
+//!
+//! The outer API of this module is the [`List`] struct. It wraps all acceptable operations on top
+//! of the aggregate linked list. All operations with the bags list should happen through this
+//! interface.
+
+use crate::Config;
+use codec::{Decode, Encode};
+use frame_election_provider_support::{VoteWeight, VoteWeightProvider};
+use frame_support::{traits::Get, DefaultNoBound};
+use sp_std::{
+	boxed::Box,
+	collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+	iter,
+	marker::PhantomData,
+};
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+	/// A duplicate id has been detected.
+	Duplicate,
+}
+
+#[cfg(test)]
+mod tests;
+
+/// Given a certain vote weight, to which bag does it belong to?
+///
+/// Bags are identified by their upper threshold; the value returned by this function is guaranteed
+/// to be a member of `T::BagThresholds`.
+///
+/// Note that even if the thresholds list does not have `VoteWeight::MAX` as its final member, this
+/// function behaves as if it does.
+pub(crate) fn notional_bag_for<T: Config>(weight: VoteWeight) -> VoteWeight {
+	let thresholds = T::BagThresholds::get();
+	let idx = thresholds.partition_point(|&threshold| weight > threshold);
+	thresholds.get(idx).copied().unwrap_or(VoteWeight::MAX)
+}
+
+/// The **ONLY** entry point of this module. All operations to the bags-list should happen through
+/// this interface. It is forbidden to access other module members directly.
+//
+// Data structure providing efficient mostly-accurate selection of the top N id by `VoteWeight`.
+//
+// It's implemented as a set of linked lists. Each linked list comprises a bag of ids of
+// arbitrary and unbounded length, all having a vote weight within a particular constant range.
+// This structure means that ids can be added and removed in `O(1)` time.
+//
+// Iteration is accomplished by chaining the iteration of each bag, from greatest to least. While
+// the users within any particular bag are sorted in an entirely arbitrary order, the overall vote
+// weight decreases as successive bags are reached. This means that it is valid to truncate
+// iteration at any desired point; only those ids in the lowest bag can be excluded. This
+// satisfies both the desire for fairness and the requirement for efficiency.
+pub struct List<T: Config>(PhantomData<T>);
+
+impl<T: Config> List<T> {
+	/// Remove all data associated with the list from storage.
+	pub(crate) fn clear() {
+		crate::CounterForListNodes::<T>::kill();
+		crate::ListBags::<T>::remove_all(None);
+		crate::ListNodes::<T>::remove_all(None);
```

Comment:
These should be bounded, not just simply call `remove_all`

Reply:
^ this combination should be illegal for any production code

Reply:
I agree with the comment overall, but to shed some light, in that PR this function is only used before `regenerate` which is called once when you initialize the list, either in the first migration or genesis. In both cases no previous data should be there and this is just acting as "better safe than sorry".

Reply:
then this shouldn't be a separate function like this. very easy for people to use this API when they shouldn't.

Or this API itself should pass in the max number to clear, and then we pass `None` in the specific scenarios we know should be okay.

Reply:
part of d8f3c1c6fac4b774dc26413f252b7bf1f85903c4 addresses this by adding a parameter to clear

Reply:
if `SortedListProvider::clear` also accept the `items: Option<u32>` then we're all happy right? why feature gate it?

Reply:
I figure if we don't use it, then it should be feature gated to avoid code bloat, but I can revert

Reply:
Prefer fixing the API rather than hiding it. Should take a `limit: Option<u32>`, and return the count of items removed.

Reply:
385138f6328c2f91ff60f53180315b27696b5e80